server/get_header.go

```diff
@@ -71,18 +76,90 @@ func (m *BoostService) getHeader(log *logrus.Entry, ua UserAgent, slot phase0.Sl
 			url := relay.GetURI(fmt.Sprintf("/eth/v1/builder/header/%d/%s/%s", slot, parentHashHex, pubkey))
 			log := log.WithField("url", url)
 
-			// Send the get bid request to the relay
-			bid := new(builderSpec.VersionedSignedBuilderBid)
-			code, err := SendHTTPRequest(context.Background(), m.httpClientGetHeader, http.MethodGet, url, ua, headers, nil, bid)
+			// A method for sending the request with a specific accept header value
+			doRequest := func(accept string) (*http.Response, error) {
+				req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, url, nil)
+				if err != nil {
+					log.WithError(err).Warn("error creating new request")
+					return nil, err
+				}
+				for key, values := range header {
+					req.Header[key] = values
+				}
+				req.Header.Set("Accept", accept)
+				return m.httpClientGetHeader.Do(req)
+			}
+
+			// Send the get bid request to the relay. Try what the client
+			// accepts (either SSZ or JSON) first. If no accept type is specified,
+			// request JSON. We cannot request SSZ if the client does not, because
+			// the appropriate Eth-Consensus-Version header value is not known.
+			var err error
+			var resp *http.Response
+			switch header.Get("Accept") {
+			case "application/octet-stream":
+				log.Debug("requesting header in SSZ")
+				resp, err = doRequest("application/octet-stream")
+				if err != nil {
+					log.WithError(err).Warn("error calling getHeader on relay")
+					return
+				}
+
+				// Check if the relay supports SSZ requests
+				if resp.StatusCode != http.StatusNotAcceptable {
```

Comment:
While this approach works I am not sure about the feasibility in practice due to time constraints, I think it would be best to use q-weighted Accept headers as [noted here](https://github.com/ethereum/builder-specs/pull/104/files#diff-9632ee7eb2afc1f79cb5a37313d0688a6518ffaf8315da0fa14761105d1faf1fR12). To handle the response body, you can look at the Content-Type header sent by server.

Reply:
Oh interesting. So this can/should be done in a single request. I'll just forward whatever the client sends & handle the response with Content-Type like you suggest.